### ISSUE TITLE
The road to razzle 4.0

### ISSUE THREAD

**Issue by fivethreeo** [jaredpalmer/razzle#1377]
What should go in razzle 4.0. I have added lots of new goodies in the canary branch.

[The branch (canary)](https://github.com/jaredpalmer/razzle/tree/canary)

[The branch documentation (canary)](https://razzle-git-canary.jared.vercel.app/)

The features I currently feel can be ready for 4.0 is:

* webpack 5 support (ready)
* new babel configuration, with ability to modify options to presets and adding presets in .babelrc so you dont need to configure everything from scratch, taken from next.js (ready but needs more eyes)
* new razzle plugin system with more hooks, modifyPaths, modifyOptions to modify a new options key that comes with defaults in razzle.config.js, modifyWebpackOptions to modify options that will be used in the webpack configuration, modifyWebpackConfig same as what plugins did before, modifyJestConfig to modify the jest configuration. Same functionality will be added to razzle.config.js so plugins become highly composable. (ready but needs more eyes)
* experimental support for react-refresh (ready)
* make configuration promise based (ready)
* add razzle export to export a static site (ready)
* new way of adding externals (ready)
* built in typescript support (ready)
* splitChunks configuration in optimize (ready but needs more eyes)
* use contenthash instead of build hash (ready)

To try it now:

```bash
npx create-razzle-app@canary testing --ignore-existing
cd testing
yarn add webpack@5.1.3 --dev
yarn start
yarn build
yarn start:prod
yarn test
```

[See the docs](https://razzle-git-canary.jared.vercel.app/getting-started)

https://github.com/jaredpalmer/razzle/releases/tag/v4.0.0-canary.0
https://github.com/jaredpalmer/razzle/releases/tag/v4.0.0-canary.1
https://github.com/jaredpalmer/razzle/releases/tag/v4.0.0-canary.2

Note: expect bugs ;)

**Comment by fivethreeo** [comment]
Instructions for testing the branch:

Install requirements and clone/check out the branch

```bash
sudo apt-get install jq 
sudo npm install -g yalc
git clone razzle
cd razzle
git checkout next-awesome
```

```bash
jq '.packages | join(" ")' --raw-output lerna.json |
  xargs -I {} sh -c 'echo {}' |
  tr '[\n ]' '\n' |
  xargs -n1 -I {} bash -c '{\
    echo cd "{}" &&
    echo yalc publish &&
    echo cd .. \
  ;}'

jq '.packages | join(" ")' --raw-output lerna.json |
  xargs -I {} sh -c 'echo {}' |
  tr '[\n ]' '\n' |
  xargs -I {} basename {} |
  xargs -n1 -I {} bash -c '\
    echo yalc add {}'
```

Load a example 

```bash
node -e 'require("./test/fixtures/util").setupStageWithExample("example", "basic", symlink=false, yarnlink=false, install=true, test=false);'
cd example
```

Run the yalc add commands in this directory.

Then

```bash
yarn build
yarn start
yarn test
yarn export # if enabled, with-react-router is a example with this 
```

Will add some more docs on toggling experimental features at 
https://razzle-git-next-awesome.jared.vercel.app/

**Comment by pseudo-su** [comment]
Are you able to publish beta/prerelease versions for testing? EG how [`react-router` does](https://github.com/ReactTraining/react-router/releases/tag/v6.0.0-beta.0)?

**Comment by fivethreeo** [comment]
I am working on doing a canary release. Was hoping someone could try it out before that. But I think I will just do a canary release now so it is easier to try out.

**Comment by fivethreeo** [comment]
This will be a 3.3 canary(dodo?). Then I will make the changes necessary to do a 4.0 canary with breaking changes and webpack 5 to the next branch. Once 3.3 gets a stable release. I will do a canary 4.0.

**Comment by fivethreeo** [comment]
Did a canary release of 3.3.0 now :)

https://github.com/jaredpalmer/razzle/releases/tag/v3.3.0-canary.0
https://github.com/jaredpalmer/razzle/releases/tag/v3.3.0-canary.3

**Comment by fivethreeo** [comment]
I think waiting for 3.3 stable to start working on 4.0 will take too long. Think I will do parallel canary’s. Maybe a finch?

**Comment by fivethreeo** [comment]
Starting to be ready to do a 4.0.0-finch.0 release of razzle. I have some issues with testing on windows. Would be great if someone can check out the next branch and get windows tests passing.

**Comment by fivethreeo** [comment]
Released a finch release now

https://github.com/jaredpalmer/razzle/releases/tag/v4.0.0-finch.0

Note: expect bugs ;)

**Comment by heath-freenome** [comment]
How soon until this is ready? Will there be an upgrade guide?

**Comment by fivethreeo** [comment]
https://github.com/jaredpalmer/razzle/issues?q=is%3Aopen+is%3Aissue+milestone%3A4.0

**Comment by SimiS22** [comment]
can't wait to try this out 🎉

**Comment by fivethreeo** [comment]
So, razzle 4.0 is really near. I promise :) If anoyne uses the canary please try out the latest so we can find any glaring issues.

**Comment by fivethreeo** [comment]
Plan was this week, but had some publishing issues. Will try monday next week. Try canary 50 in the meantime just to make sure it is ok.

**Comment by fivethreeo** [comment]
https://www.npmjs.com/package/razzle/v/4.0.1-canary.0 - needed to bump patch

**Comment by fivethreeo** [comment]
Discovered a issue with html-webpack-plugin. Need to look at that before any 4.0.1 release. May be a easy fix with peerDeps. May still release monday, if this is the case.

**Comment by fivethreeo** [comment]
If anyone uses the canary can you try the latest  '4.0.1-canary.6' ?

Think I fixed the issue with html-webpack-plugin.

See https://razzle-git-canary-jared.vercel.app/getting-started#dependencies

Just add @4.0.1-canary.6 to any core razzle packages you install.

Or see if this works for you

```bash
npx create-razzle-app@canary --verbose --example with-typeorm-graphql with-typeorm-graphql
```

If this causes no issues for you I will release 4.0.1. Just need confirmation that it works for you and not just for me :)

**Comment by fivethreeo** [comment]
Released 4.0.1-canary.15 . Added tests for installing/building/starting examples, will retry releasing on monday.

**Comment by fivethreeo** [comment]
Razzle 4.0.1 is released!! Closing
